### PR TITLE
feat: implement interface-level registry-zookeeper service registry

### PR DESCRIPTION
**Pull request by DMwangnima**
#### Check the PR title.

- [ ] This PR title match the format: \<type\>(optional scope): \<description\>
- [ ] The description of this PR title is user-oriented and clear enough for others to understand.
- [ ] Attach the PR updating the user documentation if the current PR requires user awareness at the usage level. [User docs repo](https://github.com/cloudwego/cloudwego.github.io)

#### (Optional) More detailed description for this PR(en: English/zh: Chinese).

en:
zh(optional):

### CODE REVIEW

File: registries/common.go [hunk 1]
```diff
@@ -53,21 +81,106 @@ func (u *URL) FromString(raw string) error {
...
+		return fmt.Errorf("addr %s missing port", addr)
+	}
+	if host == "" || host == "::" {
+		ipv4, err := getLocalIPV4Address()
+		if err != nil {
+			return fmt.Errorf("get local ipv4 error, cause %s", err)
+		}
+		u.host = ipv4 + ":" + port
+	} else {
+		u.host = host + ":" + port
```
Comment (felix021) [inline registries/common.go]: Seems that IPv6 address is not dealt with properly.

```
SplitHostPort splits a network address of the form "host:port", "host%zone:port", "[host]:port" or "[host%zone]:port" into host or host%zone and port.

A literal IPv6 address in hostport must be enclosed in square brackets, as in "[::1]:80", "[::1%lo0]:80".
```
https://pkg.go.dev/net#SplitHostPort

File: registries/common.go [hunk 2]
```diff
@@ -53,21 +81,106 @@ func (u *URL) FromString(raw string) error {
...
+		return err
+	}
+	finalParams := make(url.Values)
+	for key, val := range params {
+		if dubboKey, ok := outboundDubboRegistryKeysMapping[key]; ok {
+			finalParams.Set(dubboKey, val)
+		}
+		if key == DubboServiceInterfaceKey {
+			u.interfaceName = val
+			missingInterfaceFlag = true
```
Comment (felix021) [inline registries/common.go]: Why not return the error directly?

Comment (felix021) [reply registries/common.go]: And is it wrongly implemented? missingInterfaceFlag should be initialized to be true and if there's a key `DubboServiceInterfaceKey` then we set it to false?

Comment (felix021) [reply registries/common.go]: Unit tests are helpful.

File: registries/zookeeper/registry/registry.go [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+	conn, eventChan, err := zk.Connect(o.Servers, o.SessionTimeout)
+	if err != nil {
+		return nil, err
+	}
+	if o.Username != "" && o.Password != "" {
+		if err := conn.AddAuth("digest", []byte(fmt.Sprintf("%s:%s", o.Username, o.Password))); err != nil {
+			return nil, err
+		}
+	}
+	halfTimeout := o.SessionTimeout / 2
```
Comment (felix021) [inline registries/zookeeper/registry/registry.go:58]: Why `/2`? Please add a simple comment explaining this.

File: registries/zookeeper/registry/registry.go [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+	}
+	u := new(registries.URL)
+	if err := u.FromInfo(info); err != nil {
+		return err
+	}
+	path := u.GetRegistryServiceKey(z.opt.RegistryGroup)
+	content := u.ToString()
+	finalPath := path + "/" + content
+	z.mu.Lock()
+	defer z.mu.Unlock()
```
Comment (felix021) [inline registries/zookeeper/registry/registry.go]: Is the critical section of this lock too big? (line 170 && line 173)

File: registries/zookeeper/registry/registry.go [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+			_, err = z.conn.Create(path, content, flag, zk.WorldACL(zk.PermAll))
+		}
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (z *zookeeperRegistry) keepalive(ctx context.Context, path string, content []byte) {
```
Comment (felix021) [inline registries/zookeeper/registry/registry.go:135]: Just to be sure: If the server process exits unexpectedly (not calling Deregister), will the entry of this instance in zookeeper be removed automatically (after a while)?

File: registries/zookeeper/resolver/resolver.go [hunk 6]
```diff
@@ -105,3 +108,22 @@ func (z *zookeeperResolver) Diff(cacheKey string, prev, next discovery.Result) (
 func (z *zookeeperResolver) Name() string {
 	return z.uniqueName
 }
+
```
Comment (felix021) [inline registries/zookeeper/resolver/resolver.go:112]: This function is hard to understand, please add some comment and an example input/output

Comment (felix021) [reply registries/zookeeper/resolver/resolver.go:112]: Also better covered with unit tests

File: registries/common.go (hunk 1 above)
Comment (DMwangnima) [reply registries/common.go]: Ack.

File: registries/common.go (hunk 2 above)
Comment (DMwangnima) [reply registries/common.go]: I would add unit tests to fix it.

File: registries/zookeeper/registry/registry.go (hunk 4 above)
Comment (DMwangnima) [reply registries/zookeeper/registry/registry.go]: Yes. It should be released after processing ```canceler``` map.

File: registries/zookeeper/registry/registry.go (hunk 5 above)
Comment (DMwangnima) [reply registries/zookeeper/registry/registry.go:135]: Yes, the leaf would be ephemeral node but its ancestor would be persistent nodes. This is consistent with the implementation of dubbo.

File: registries/zookeeper/resolver/resolver.go (hunk 6 above)
Comment (DMwangnima) [reply registries/zookeeper/resolver/resolver.go:112]: Ack.

File: registries/zookeeper/registry/registry.go (hunk 3 above)
Comment (DMwangnima) [reply registries/zookeeper/registry/registry.go:58]: This is just a value in the middle range. It should not exceed sessionTimeout and should not be too small. I would add a comment to explain it.

**Comment by felix021** [comment]
Add a comment to pause bot notice.

Notify me after new push.

File: registries/zookeeper/registry/registry.go [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+	}
+	path := u.GetRegistryServiceKey(z.opt.RegistryGroup)
+	content := u.ToString()
+	finalPath := path + "/" + content
+	z.mu.Lock()
+	cancel, ok := z.canceler[finalPath]
+	if ok {
+		cancel()
+		delete(z.canceler, finalPath)
+		z.mu.Unlock()
```
Comment (felix021) [inline registries/zookeeper/registry/registry.go]: It's dangerous: what if `ok == false`?

Comment (felix021) [reply registries/zookeeper/registry/registry.go]: A more elegant way is to encapsulate `mu` with `canceler` to precisely control the critical zone

**Review by felix021 — APPROVED** [review]

**Pull request merged**